azure-ts-serverless-url-shortener-global/index.ts

```diff
@@ -0,0 +1,120 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as azure from "@pulumi/azure";
+import { Container } from "@azure/cosmos";
+import { getContainer } from "./cosmosclient";
+
+// Read a list of target locations from the config file:
+// Expecting a comma-separated list, e.g., "westus,eastus,westeurope"
+const config = new pulumi.Config();
+const locations = config.require("locations").split(',');
+// The first location is considered primary
+const primaryLocation = locations[0];
+
+let resourceGroup = new azure.core.ResourceGroup("urlshort-rg", {
+    location: primaryLocation,
+});
+
+// Cosmos DB with a single write region (primary location) and multiple read replicas
+let cosmosdb = new azure.cosmosdb.Account("url-cosmos", {
+    resourceGroupName: resourceGroup.name,
+    location: primaryLocation,
+    geoLocations: locations.map((location, priority) => ({
+        location: location,
+        failoverPriority: priority,
+    })),
+    offerType: "Standard",
+    consistencyPolicy: {
+        consistencyLevel: "Session",
+        maxIntervalInSeconds: 5,
+        maxStalenessPrefix: 100,
+    },
+});
+
+// Traffic Manager as a global HTTP endpoint
+const profile = new azure.trafficmanager.Profile("urlshort-tm", {
+    resourceGroupName: resourceGroup.name,
+    trafficRoutingMethod: 'Performance',
+    dnsConfigs: [{
+        relativeName: "urlshort-tm",
+        ttl: 60
+    }],
+    monitorConfigs: [{
+        protocol: 'HTTP',
+        port: 80,
+        path: '/api/ping',
+    }]
+});
+
+// Azure Function to accept new URL shortcodes and save to Cosmos DB
+const fn = new azure.appservice.HttpEventSubscription("urlshort-add", {
+    resourceGroup,
+    location: primaryLocation,
+    methods: ["POST"],
+    callbackFactory: () => {
+        const endpoint = cosmosdb.endpoint.get();
+        const masterKey = cosmosdb.primaryMasterKey.get();
+
+        let container: Container;
+        return async (_, request: azure.appservice.HttpRequest) => {
+            container = container || await getContainer(endpoint, masterKey, primaryLocation);
+
+            await container.items.create(request.body);
+            return { status: 200, body: 'Short URL saved' };
+        };
+    }
+});
+export const addEndpoint = fn.url;
+
+for (const location of locations) {
+
+    // URL redirection function - one per region
+    const fn = new azure.appservice.HttpEventSubscription(`urlshort-fa-${location}`, {
+        resourceGroup,
+        location,
+        route: "{key}",
+        callbackFactory: () => {
+            const endpoint = cosmosdb.endpoint.get();
+            const masterKey = cosmosdb.primaryMasterKey.get();
+
+            let container: Container;
+            return async (_, request: azure.appservice.HttpRequest) => {
+                container = container || await getContainer(endpoint, masterKey, location);
```

Comment:
is this line expected to be expensive?  if so, callbackFactory is appropriate to use here.  If it's not expensive, i would just use `callback: async () ... ` and just compute this on demand here.

Reply:
It is sort of expensive because it checks if the database and the container exist and creates them otherwise. Ideally, I would put this initialization outside of the callback, but `await` is not allowed there.
For this example, we can change back to `callback` to simplify, but in my blog I'm making a case for fast response time, and those call screw the latency.

Reply:
Well, ideally ideally, the database and container would be created by ARM/Pulumi, but they aren't managed by ARM (yet).